### DISCUSSION TITLE
Using an accurate benchmark tool

### DISCUSSION THREAD

**Discussion by waghanza** [the-benchmarker/web-frameworks#8088] [Show and tell]
We are currently using wrk. This tool is great, no doubt about that. However, the purpose of such a tool is to check how a framework could perform at its best (grab the maximum performance it could give).

This project aims to reflect how frameworks could perform in real world scenarios (for example, using constant throughput). For this reason, I'm considering switching this tool.

Candidates are :
- https://github.com/tsenart/vegeta
- https://github.com/tag1consulting/goose
- https://github.com/codesenberg/bombardier
- https://github.com/sharkdp/hyperfine

Goose seems accurate for me. Any suggestions @dominikzogg @jeremyevans @panesofglass @jodosha @timriley @ahopkins @appleboy @yusukebe @p8 @mcollina @fundon @ohler55 @sy-records @enghitalo @lukaszlenart ?

PS : The idea is to change this tool on Q2 2025

PSS : Sorry to ping some individuals, I do not know how to ping the whole team at once 😇

**Comment by waghanza** [comment]
Sorry : not hyperfine (more to benchmark cli tools)

**Comment by fundon** [comment]
https://github.com/hatoo/oha
It is also a good testing tool.

**Reply by waghanza** [reply]
Seems more like a PoC for me https://github.com/hatoo/oha/issues/617

But to be honest, it's promising

**Comment by appleboy** [comment]
Yes, `tsenart/vegeta` is a versatile HTTP load testing tool that can handle real-world scenarios, including constant throughput, which is what you're looking for. Vegeta is designed to be both a command-line utility and a library, making it flexible for various use cases.

## Key Features of Vegeta

1. Constant Throughput: Vegeta can maintain a constant request rate, which is useful for simulating real-world traffic patterns.
2. Ease of Use: It has a simple command-line interface and can be easily integrated into scripts and CI/CD pipelines.
3. Detailed Reporting: Vegeta provides detailed metrics and can generate reports in various formats, including JSON and HTML.
4. Flexibility: It supports multiple attack modes, including rate-limited and burst modes, and can handle complex scenarios with custom headers, bodies, and methods.

**Reply by waghanza** [reply]
I'm concerned about garbage collecting that can mess up results. Also concurrency model (async / await) seems more hard to learn, but more accurate in this context, I mean vs rust (goose)

**Reply by waghanza** [reply]
Also, how can vegeta handle constant throughput ? I mean, how can we make an equal sign between constant throughput and a real world scenario, there is no control about throughout

**Reply by waghanza** [reply]
Vegeta use 50rps per default, how can we define the rate to use facing the scenario we use ?

**Reply by waghanza** [reply]
I have tested `vegeta`  (without rate limit), just to compare to `wrk` (since it has not rate limit) , and I can not explain result I've found in https://github.com/tsenart/vegeta/issues/720

**Reply by waghanza** [reply]
How can I use constant throughput in vegeta ?

How can I use rate param in vegeta ? Nad this it really accurate in benchmark context @appleboy 

cc @akyoto

**Reply by akyoto** [reply]
> How can I use constant throughput [...] Nad this it really accurate in benchmark context

Constant throughput is useful in private benchmarks at a company with very clear constraints.
[Constant throughput should be avoided for public benchmarks.](https://urbach.dev/realistic-benchmarks)

**Comment by mcollina** [comment]
I think wrk is ok. wrk2 is better ;).

**Reply by waghanza** [reply]
I do not understand how using constant throughput could be helpful.

I mean, with constant throughput approach, we must know the rate we should use. With wrk2, the rate used (by default) is 1000, but how can we know that 1000 is accurate for our scenario ? Or define any rate for it ?

**Reply by waghanza** [reply]
wrk2 is not yet compilable for arm

**Comment by yusukebe** [comment]
I prefer to [bombardier](https://github.com/codesenberg/bombardier), though I don't have any concrete reason. My favorite benchmark by @SaltyAom  is using it:

https://github.com/SaltyAom/bun-http-framework-benchmark

**Reply by waghanza** [reply]
I there a way to enable any rate limit with it @yusukebe @SaltyAom ?

This feature is important, imho, to make more realistic results

**Reply by alanmsant2** [reply]
I use rewrk (Rust) [link](https://github.com/lnx-search/rewrk)

**Reply by waghanza** [reply]
Any rationales behind ?

**Reply by alanmsant2** [reply]
In my perception, I found the results more accurate with Rewrk

**Comment by panesofglass** [comment]
I've only used bombardier, and I thought it did a good job. I'll have to try the others, but I cannot offer any good comparisons.

**Comment by cyrusmsk** [comment]
In "Clients" folder some comparison of generator tools as well
https://github.com/dpbench/dpbench

**Comment by MDA2AV** [comment]
higher throughput than any other load generator and precise latency values

https://github.com/MDA2AV/gcannon

**Reply by waghanza** [reply]
When I see https://www.http-arena.com, I can see that actix is ~2.5 better than rails / flask -> https://www.http-arena.com/leaderboard/#v=composite&res=mem&lang=Python%2CRuby%2CRust&q=actix%2Cflask%2Crails

We have totally other results -> https://web-frameworks-benchmark.netlify.app/compare?f=actix,rails,flask

**Reply by MDA2AV** [reply]
Can't compare composite score results with a single test, if you compare the closest test (baseline) httparena has with web frameworks benchmark then actix is 53x faster than rails (2.9M to 55K), naturally on heavier load tests like json, database access etc test this gap closes.

k6 isn't an option for me, it can't saturate servers because it hits ceiling before the server being tested and basically measures nothing or completely irrelevant results. It is not a bad tool but completely mismatched for a benchmark platform

**Reply by cyrusmsk** [reply]
both solutions looks like vibe-coded in couple of weeks.. mostly by 1 author.
AI quality of code doesn't have such level of trust in my opinion yet.
How these solutions will be able to handle complexity - will see after some time

Results in the leaderboard look very suspicious

**Reply by waghanza** [reply]
you mean oha and gcannon @cyrusmsk ?

**Reply by cyrusmsk** [reply]
gcannon
oha is in the development for several years and has plenty of contributors

**Reply by waghanza** [reply]
> actix is 53x faster than rails (2.9M to 55K)

how do you check that @MDA2AV ?

**Reply by waghanza** [reply]
> both solutions looks like vibe-coded in couple of weeks.

you mean gcannon and httparena @cyrusmsk  ?

**Reply by waghanza** [reply]
ah I see ... and rage 364K

very different that what we have https://web-frameworks-benchmark.netlify.app/compare?f=actix,rails,flask,rage

**Reply by MDA2AV** [reply]
> ah I see ... and rage 364K
> 
> very different that what we have https://web-frameworks-benchmark.netlify.app/compare?f=actix,rails,flask,rage

It's running on very different hardware, also different test types, I wouldn't expect results to be any similar

**Reply by MDA2AV** [reply]
oha is too slow, can't saturate servers. The only long lived option is wrk but it doesn't support reconnects, gcannon fixes that, it is a port from a C# project into C but it works fine, it isn't a vibe coded project but some parts of it were built with AI assistance to speed it up, not the core io_uring mechanism.

httparena doesn't use gcannon only though, it uses wrk, h2load, h2load-h3 and ghz for load generators, also gcannon results correlate to wrk with 5-10% performance increase on all tests

**Reply by Kaliumhexacyanoferrat** [reply]
> Results in the leaderboard look very suspicious

Results seem to have sorted - they refactored the tests a few times but now they seem quite balanced. And the actual run on real potent hardware, similar to TFB.

**Reply by waghanza** [reply]
> > ah I see ... and rage 364K
> > very different that what we have https://web-frameworks-benchmark.netlify.app/compare?f=actix,rails,flask,rage
> 
> It's running on very different hardware, also different test types, I wouldn't expect results to be any similar

What is more important for me is the ratio (rails is by X less performant than actix ...) not absolute values @MDA2AV

**Reply by MDA2AV** [reply]
Yes, but the tests would have to be similar, for example if you compare the CRUD test with a simple hello world test, the CRUD test will exercise the frameworks' threadpool, database driver or async/await performance, while the simple hello world test is gated by the framework's I/O and parser, even inside parser some frameworks are faster with chunked encoding requests and different parts of the parsing logic. Even if the tests are similar, by testing on a machine with 128 cores the scaling might not be the same as on a machine with 16, I've seen many frameworks that can't scale beyond 12 cores

**Reply by p8** [reply]
If you look at the TFB results for Actix and Rails...

On the plaintext test Actix is about 50 times faster:
<img width="1174" height="171" alt="image" src="https://github.com/user-attachments/assets/7f0e7081-c9d0-4881-9241-da12538b109a" />

https://www.techempower.com/benchmarks/#section=data-r22&f=zhb20v-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-yyku7z-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-1ekf&test=plaintext

On the JSON test Actix is about 13 times faster:
<img width="1175" height="172" alt="image" src="https://github.com/user-attachments/assets/c15a5e10-5919-4e35-8726-23f0c337cbe7" />

https://www.techempower.com/benchmarks/#section=data-r22&f=zhb20v-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-yyku7z-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-1ekf&test=json

On the Multi query Actix is 4 times faster:
<img width="1180" height="259" alt="image" src="https://github.com/user-attachments/assets/9dcf80d7-4f38-4a7f-8f09-fb5f565f2e8c" />


https://www.techempower.com/benchmarks/#section=data-r22&f=zhb20v-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-yyku7z-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-1ekf&test=query

**Reply by waghanza** [reply]
ℹ️ 

- TFB is archived (at least the repo)
- Rust is 158 -> https://github.com/TechEmpower/FrameworkBenchmarks/blob/523534bb61450e3522d775a749ad060753e26e3a/frameworks/Rust/actix/actix-http.dockerfile#L1
- Ruby is 3.4 -> https://github.com/TechEmpower/FrameworkBenchmarks/blob/523534bb61450e3522d775a749ad060753e26e3a/frameworks/Ruby/rails/rails.dockerfile#L1
- Rails is 8.0 -> https://github.com/TechEmpower/FrameworkBenchmarks/blob/523534bb61450e3522d775a749ad060753e26e3a/frameworks/Ruby/rails/Gemfile#L3

**Comment by waghanza** [comment]
Maybe [k6](https://k6.io) is more accurate for what we do

**Reply by floatdrop** [reply]
Tried it in cloud, but k6 is a resource hog and needs a lot of cpu.

**Reply by waghanza** [reply]
Actually `k6` eats some CPU, so benchmark figures will not be relevant. However :
- benchmark figure is not a good indicator -> relative figure (percents) will be more accurate for us
- in a near future, client will be separated from servers, so CPU eating in client side will not be a problem anymore

**Reply by floatdrop** [reply]
> in a near future, client will be separated from servers, so CPU eating in client side will not be a problem anymore

Well, it will affect the cost of benchmark run (I was forced to spin up multiple k6 instances to achieve desired throughput) and it was biting my wallet.

**Comment by floatdrop** [comment]
I have published a [zrk](https://zoxy.io/zrk/) – wrk2 port to Zig with machine readable reports and TUI. You can check it out.

**Reply by floatdrop** [reply]
@waghanza better late, than never: little follow up – https://github.com/the-benchmarker/web-frameworks/pull/9668
